alerts/plugins/ip_source_enrichment.py

```diff
@@ -0,0 +1,146 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# Copyright (c) 2014 Mozilla Corporation
+
+from operator import add
+import os
+import re
+
+import netaddr
+
+
+CONFIG_FILE = os.path.join(
+    os.path.dirname(__file__),
+    'ip_source_enrichment.json')
+
+
+def _find_ip_addresses(string):
+    '''List all of the IPv4 and IPv6 addresses found in a string.'''
+
+    ipv4_rx = '(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)'
+    ipv6_rx = '(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))'
+
+    ipv4 = re.findall(ipv4_rx, string)
+    ipv6 = map(
+        lambda match: match[0] if isinstance(match, tuple) else match,
+        re.findall(ipv6_rx, string))
+
+    return ipv4 + ipv6
+
+
+def enrich(alert, known_ips):
+    '''Add information to alerts containing IP addresses that describes
+    the source location of the IP address if it can be determined based
+    on a configured mapping.
+    '''
+
+    def find_ips(value):
+        if isinstance(value, str):
+            return _find_ip_addresses(value)
+
+        if isinstance(value, list) or isinstance(value, tuple):
+            found = [find_ips(item) for item in value]
+            return reduce(add, found, [])
+
+        if isinstance(value, dict):
+            found = [find_ips(item) for item in value.values()]
+            return reduce(add, found, [])
+
+        return []
+
+
+    def ip_in_range(ip):
+        return lambda known: netaddr.IPAddress(ip) in netaddr.IPSet([known['range']])
+
+
+    ips = find_ips(alert)
+
+    alert = alert.copy()
+        
+    alert['details']['sites'] = []
+
+    for ip in set(ips):
+        matching_descriptions = filter(ip_in_range(ip), known_ips)
+
+        for desc in matching_descriptions:
+            enriched = desc['format'].format(ip, desc['site'])
+            
+            alert['summary'] += '; ' + enriched
+
+            alert['details']['sites'].append({
+                'ip': ip,
+                'site': desc['site'],
+            })
+
+    return alert
+
+
+def _load_config(file_path):
```

Comment:
Well, the _point_ of the function is to load the alert plugin's configuration. I hoped that would be obvious by:

1. The name
2. Its usage context and
3. Its doc string

To answer the question you didn't ask:

I left function in its sort of half-baked state because I wasn't sure what the recommended way to open up a config file is.  Normally this would be a basic thing but I know the alert plugins are run on workers and might have a different base path from another component.

Would an implementation as simple as

```py
try:
  f = open(file_path)
  return json.load(f)
except IOError:
  return { 'known': [] }
```

be okay?

Reply:
Yeah, that looks good from a config loading perspective.